core/src/main/java/com/linecorp/armeria/common/logging/PreviewSpec.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import java.nio.charset.Charset;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.RequestContext;
+
+import io.netty.buffer.ByteBuf;
+
+final class PreviewSpec {
+
+    private final BiPredicate<? super RequestContext, ? super HttpHeaders> predicate;
+    private final PreviewMode mode;
+    @Nullable
+    private final Charset defaultCharset;
+    @Nullable
+    private final BiFunction<? super HttpHeaders, ? super ByteBuf, String> producer;
+
+    PreviewSpec(BiPredicate<? super RequestContext, ? super HttpHeaders> predicate,
+                PreviewMode mode, @Nullable Charset defaultCharset,
+                @Nullable BiFunction<? super HttpHeaders, ? super ByteBuf, String> producer) {
+        this.predicate = predicate;
+        this.mode = mode;
+        this.defaultCharset = defaultCharset;
+        this.producer = producer;
+    }
+
+    BiPredicate<? super RequestContext, ? super HttpHeaders> predicate() {
+        return predicate;
+    }
+
+    PreviewMode mode() {
+        return mode;
+    }
+
+    @Nullable
+    Charset defaultCharset() {
```

Comment:
Does a preview spec also need a default charset? I'm having trouble understanding how this is used.

Reply:
Maybe it's too much to have different default charsets for different media types?
I think I can just remove this and use one default charset for every text preview when the charset is not specified in the content type. Thanks!